src/azul/indexer/__init__.py

```diff
@@ -482,6 +509,36 @@ class SourcedBundleFQID(BundleFQID, Generic[SOURCE_REF]):
 
     source: SOURCE_REF
 
+    def __init_subclass__(cls, **kwargs):
+        """
+        >>> @attrs.frozen(kw_only=True)
+        ... class FooBundleFQID(SourcedBundleFQID):
+        ...     foo: str
+        Traceback (most recent call last):
+        ...
+        AssertionError: <class 'azul.indexer.FooBundleFQID'>
+
+        >>> @attrs.frozen(kw_only=True, eq=False)
+        ... class FooBundleFQID(SourcedBundleFQID):
+        ...     foo: str
+        """
+        # Ideally this would go in the parent class, but it's pushed down here
+        # to facilitate the assertion in :meth:`__eq__`
+        super().__init_subclass__(**kwargs)
+        assert cls.__eq__ is SourcedBundleFQID.__eq__, cls
+        assert cls.__hash__ is SourcedBundleFQID.__hash__, cls
+
+    def __hash__(self):
+        # Since we define a custom `__eq__` we're forced to also define a custom
+        # `__hash__`, even though both implementations just call super()
+        return super().__hash__()
+
+    def __eq__(self, other: 'BundleFQID') -> bool:
```

Comment:
I think it is perfectly fine to move this functionality to `BundleFQID.__eq__` and use an `isinstance` check on the `self` argument to guard it. Then you can pull up the `__init__subclass__`  method.